🎉 New Source: Airtable

tuanchris opened the pull request:
## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

## How
- Create new source for Airtable 
- https://github.com/airbytehq/hacktoberfest-2021/issues/56
- https://github.com/airbytehq/airbyte/issues/1468
- Integration tests passed 
<img width="1126" alt="image" src="https://user-images.githubusercontent.com/52090179/140335402-15b7331b-6599-4238-8fc0-b02e58364101.png">
- Unit test passed 
<img width="1117" alt="image" src="https://user-images.githubusercontent.com/52090179/140335564-ac435988-74f3-41b7-8345-e23fea659fe1.png">
- Basic normalization to BigQuery succeeded <img width="1117" alt="image" src="https://user-images.githubusercontent.com/52090179/140380998-5e182db9-2436-49b8-9e0a-fc7bde75b552.png">

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>


<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

alafanechere commented on airbyte-integrations/connectors/source-airtable/source_airtable/source.py:
```diff
@@ -0,0 +1,131 @@
+            try:
+                response = requests.get(url, headers=auth)
+                response.raise_for_status()
+            except requests.exceptions.HTTPError as e:
```
Do you think you could implement a bit more complex error handling, you might be able to differentiate wrong credentials from non existing table name thanks to HTTP statuses and it would greatly help users debugging their source set up.

tuanchris replied:
added more user friendly error messages

alafanechere commented on airbyte-integrations/connectors/source-airtable/source_airtable/source.py:
```diff
@@ -0,0 +1,131 @@
+            }
+            data = record.get("fields", {})
+            # Convert all values to string
+            data = {key: str(value) for key, value in data.items()}
```
Please checkout our [Type Transformation](https://docs.airbyte.io/connector-development/cdk-python/schemas#type-transformation) documentation. You could delegate this kind of transformation to a `TypeTransformer`

tuanchris replied:
I had this before: `transformer = TypeTransformer(TransformConfig.CustomSchemaNormalization)`
![image](https://user-images.githubusercontent.com/52090179/141227119-8bcac6e1-0c15-4d17-adcb-8cbae5167c1c.png)
But still got this error running integration tests.

alafanechere commented on airbyte-integrations/connectors/source-airtable/source_airtable/source.py:
```diff
@@ -0,0 +1,131 @@
+        auth = TokenAuthenticator(token=config["api_key"])
+        streams = []
+        for table_name in config["tables"]:
+            kwargs = {"base_id": config["base_id"], "table_name": table_name, "authenticator": auth, "schema": self.schemas.get(table_name)}
```
I'm not really comfortable with the fact  that `streams` function depends of a previous `discover` call to set `self.schemas`.

tuanchris replied:
I have refactored and used helper functions for this

alafanechere reviewed (changes requested):
Hi @tuanchris , we really appreciate your contribution, thank you!
I did a first quick review of the core of your code and made my first comments / suggestions. 
May I ask you to also complete what's required in the pre-merge checklist? 
I would also appreciate some unit testing of your source and stream classes functions.

tuanchris commented on airbyte-integrations/connectors/source-airtable/source_airtable/source.py:
```diff
@@ -0,0 +1,131 @@
+        processed_records = []
+        for record in records:
+            processed_record = {
+                "_airtable_id": record.get("id"),
+                "_airtable_created_time": record.get("createdTime"),
```
Well, `id` and `createdTime` are very common column names. These columns can be overwritten if for example there is an id column in the table.

tuanchris commented on airbyte-integrations/connectors/source-airtable/source_airtable/source.py:
```diff
@@ -0,0 +1,131 @@
+        return True, None
+
+    def discover(self, logger: AirbyteLogger, config) -> AirbyteCatalog:
+        streams = []
```
Great! I have refactored some of the code out to Helper functions. Airtable already takes care of duplicated fields though. 
![image](https://user-images.githubusercontent.com/52090179/141225077-739f449b-cc7c-4d6a-a632-c991ebb5889c.png)

tuanchris commented:
@alafanechere I have added unit tests and integration tests and they have passed as follow:
![image](https://user-images.githubusercontent.com/52090179/141332149-78a360a6-56b5-44f1-89f0-8d2a12307f70.png)
![image](https://user-images.githubusercontent.com/52090179/141332368-0747ab92-24d3-4bfb-ba50-9bc36020caa9.png)

marcosmarxm commented:
Hello @tuanchris ! Sorry to not finished your contribution before the date stipulated in the contest. All contributions made before 15-November are eligible to receive the award. We're trying to review your contribution as soon as possible.

alafanechere commented on airbyte-integrations/connectors/source-airtable/unit_tests/test_helpers.py:
```diff
@@ -0,0 +1,39 @@
+        Helpers.get_first_row(auth, base_id, table)
+
+
+def test_table_not_found(auth, base_id, table):
```
```suggestion
def test_get_first_row_table_not_found(auth, base_id, table):
```

alafanechere commented on airbyte-integrations/connectors/source-airtable/unit_tests/test_helpers.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def test_get_airbyte_stream(table, expected_json_schema):
+    Helpers.get_airbyte_stream(table, expected_json_schema)
```
I think some assertions are missing here.

alafanechere commented on airbyte-integrations/connectors/source-airtable/unit_tests/test_source.py:
```diff
@@ -0,0 +1,36 @@
+def test_discover(test_config):
+    source = SourceAirtable()
+    logger_mock = MagicMock()
+    response = source.discover(logger_mock, test_config)
```
```suggestion
    airbyte_catalog = source.discover(logger_mock, test_config)
```

alafanechere replied:
I think you could add more assertions: e.g check the streams in the catalog match the tables defined in the config, or mock helpers function and check they've been called etc.

alafanechere commented on airbyte-integrations/connectors/source-airtable/unit_tests/test_source.py:
```diff
@@ -0,0 +1,36 @@
+@patch("requests.get")
+def test_streams(test_config):
+    source = SourceAirtable()
+    source.streams(test_config)
```
More assertions can be made on the `source.streams` return value.

alafanechere reviewed (changes requested):
Thank you very much for the changes you made. This is going in the right direction 🙏 ! I've some additional comments / suggestions that you can, of course, feel free to challenge. 
I succeeded in running your integrations / unit tests.
Do no hesitate to ping me here or on slack if you want me to review again.

tuanchris commented:
Hi @alafanechere, I have addressed all the comments. Please have another look! 
Unit tests passed
![image](https://user-images.githubusercontent.com/52090179/142390297-0d874331-7e3c-42e5-ab6e-12b7aba324f4.png)
Acceptance tests passed
![image](https://user-images.githubusercontent.com/52090179/142390358-585d0797-f3ab-4e13-bed3-6ce1b1d1d192.png)

alafanechere commented:
Thanks @tuanchris ! I'll check this out before end of the week.

alafanechere reviewed (approved):
@tuanchris I really appreciate the change you've made! 
I allowed myself to remove an unused import spotted by Flake. I ran the acceptance test successfully 👍 
Looks good to me, I'm now opening a side branch in order to run our private CI on it and then we'll be ready to merge 👏

tuanchris commented:
Thanks, @alafanechere for the detailed and throughout reviews. I actually picked up a thing or two about testing! 🚀
